### PR TITLE
CBQI BV choice expressions

### PR DESCRIPTION
**Pull request by ajreynol**
Updates counterexample-guided quantifier instantiation for bit-vectors to use choice expressions. Updates in this pull request: 
- Eliminates the hacks that introduce explicit Skolem functions and replaces them with choice expressions,
- Infrastructure for getBoundVariable in CegInstantiator for getting canonical bound variables for use in choice expressions. This required minor interface changes.
- Improve documentation for CegInstantiator,
- Update CegInstantiator to use more optimized data structures for computing ineligible terms and free variables,
- Fixes a bug in CegInstantiator that led to instantiations not being cleaned up after post-processing,
- Fixes a bug in TermFormulaRemoval,
- Adds several regressions that failed intermediate commits of this pull request.

### CODE REVIEW

**Comment by ajreynol** [comment]
Also, this will fix #1172.

File: src/theory/quantifiers/bv_inverter.cpp [hunk 1]
```diff
@@ -53,20 +53,8 @@ Node BvInverter::getSolveVariable(TypeNode tn)
...
-      Assert(tn.getNumChildren() == 2);
-      Assert(tn[0].isBoolean());
-      ss << "slv_f";
-    }
-    else
-    {
-      ss << "slv";
-    }
+    ss << "slv";
     Node k = NodeManager::currentNM()->mkSkolem(ss.str(), tn);
```
Comment (mpreiner) [inline src/theory/quantifiers/bv_inverter.cpp]: You can pass "slv" directly to `mkSkolem()`, then `ss` is not needed.

File: src/theory/quantifiers/ceg_instantiator.cpp [hunk 2]
```diff
@@ -46,25 +46,42 @@ CegInstantiator::~CegInstantiator() {
...
-        d_inelig[n] = true;
+        d_inelig.insert(n);
       }
-      for( std::map< Node, bool >::iterator it = d_prog_var[n[i]].begin(); it != d_prog_var[n[i]].end(); ++it ){
-        d_prog_var[n][it->first] = true;
-      }
-      //selectors applied to program variables are also variables
-      if( n.getKind()==APPLY_SELECTOR_TOTAL && d_prog_var[n].find( n[0] )!=d_prog_var[n].end() ){
-        d_prog_var[n][n] = true;
+      // all variables in child are cotained in this
```
Comment (aniemetz) [inline src/theory/quantifiers/ceg_instantiator.cpp]: coNtained

File: src/theory/quantifiers/ceg_instantiator.h [hunk 3]
```diff
@@ -134,43 +138,83 @@ class SolvedForm {
...
+   * ineligible for instantiation.
+    */
+  std::unordered_set<Node, NodeHashFunction> d_inelig;
+  /** current assertions per theory */
+  std::map<TheoryId, std::vector<Node> > d_curr_asserts;
+  /** map from representatives to the terms in their equivalence class */
+  std::map<Node, std::vector<Node> > d_curr_eqc;
+  /** map from types to representatives of that type */
+  std::map<TypeNode, std::vector<Node> > d_curr_type_eqc;
+  // auxiliary variables
```
Comment (aniemetz) [inline src/theory/quantifiers/ceg_instantiator.h]: Maybe use consistent comment style? (/**/ in this case, here and below)

File: src/theory/quantifiers/ceg_instantiator.cpp [hunk 4]
```diff
@@ -46,25 +46,42 @@ CegInstantiator::~CegInstantiator() {
...
+      for (std::unordered_set<Node, NodeHashFunction>::iterator it =
+               d_prog_var[n[i]].begin();
+           it != d_prog_var[n[i]].end();
+           ++it)
+      {
+        d_prog_var[n].insert(*it);
```
Comment (mpreiner) [inline src/theory/quantifiers/ceg_instantiator.cpp]: Can you use
```
d_prog_var[n] = std::unordered_set<Node, NodeHashFunction>(d_prog_var[n[i]].begin(), d_prog_var[n[i]].end())
```
here?

File: src/theory/quantifiers/ceg_t_instantiator.cpp [hunk 5]
```diff
@@ -1180,10 +1165,34 @@ Node BvInstantiator::rewriteAssertionForSolvePv( Node pv, Node lit ) {
...
+        TNode var = cur[0][0];
+        Node sbody = cur[1].substitute(var, bv);
+        // we cannot cache the results of subterms
+        // of this choice expression since we are
+        // now in the context { cur[0][0] -> bv },
+        // hence we make a separate call to
+        // rewriteAssertionForSolvePv here,
+        // where the recursion depth is the maximum
+        // depth of nested choice expressions.
+        Node rsbody = rewriteAssertionForSolvePv(ci, pv, sbody);
```
Comment (aniemetz) [inline src/theory/quantifiers/ceg_t_instantiator.cpp]: Even if the max recursion depth seems limited here, let's not introduce new recursive implementations and unrecursify rewriteAssertionForSolvePv (maybe in a separate PR).

File: src/theory/quantifiers/ceg_instantiator.cpp [hunk 6]
```diff
@@ -1010,6 +1054,10 @@ void CegInstantiator::registerCounterexampleLemma( std::vector< Node >& lems, st
   //Assert( d_vars.empty() );
   d_vars.clear();
   d_vars.insert( d_vars.end(), ce_vars.begin(), ce_vars.end() );
+  for (unsigned i = 0; i < ce_vars.size(); i++)
+  {
+    d_vars_set.insert(ce_vars[i]);
+  }
```
Comment (mpreiner) [inline src/theory/quantifiers/ceg_instantiator.cpp]: Use `d_vars_set.insert(ce_vars.begin(), ce_vars.end())` here.

File: src/theory/quantifiers/ceg_instantiator.h [hunk 7]
```diff
@@ -134,43 +138,83 @@ class SolvedForm {
...
-  std::vector< TheoryId > d_tids;
-  //literals to equalities for aux vars
-  std::map< Node, std::map< Node, Node > > d_aux_eq;
-  //the CE variables
-  std::vector< Node > d_vars;
-  //index of variables reported in instantiation
-  std::vector< unsigned > d_var_order_index;
-  //atoms of the CE lemma
+  std::vector<TheoryId> d_tids;
+  // literals to equalities for aux vars
```
Comment (mpreiner) [inline src/theory/quantifiers/ceg_instantiator.h]: Inconsistent comment style `/**` vs `//`

File: src/theory/quantifiers/ceg_instantiator.h (hunk 3 above)
Comment (ajreynol) [reply src/theory/quantifiers/ceg_instantiator.h]: I went ahead and added documentation for the functions in this block

File: src/theory/quantifiers/ceg_t_instantiator.cpp (hunk 5 above)
Comment (ajreynol) [reply src/theory/quantifiers/ceg_t_instantiator.cpp]: sounds good, it's possible to do this, just need to have two levels to the context in this function.

File: src/theory/quantifiers/bv_inverter.cpp (hunk 1 above)
Comment (ajreynol) [reply src/theory/quantifiers/bv_inverter.cpp]: fixed

File: src/theory/quantifiers/ceg_instantiator.cpp (hunk 4 above)
Comment (ajreynol) [reply src/theory/quantifiers/ceg_instantiator.cpp]: This needs to append, not overwrite.  Does d_prog_var[n].insert(d_prog_var[n[i]].begin(), d_prog_var[n[i]].end()) work in this case?

Comment (mpreiner) [reply src/theory/quantifiers/ceg_instantiator.cpp]: Yes this should work.

**Comment by ajreynol** [comment]
I've addressed the reviews in my latest commit.

**Pull request merged**